### PR TITLE
refactor: Introduce context cause to our code

### PR DESCRIPTION
**Pull request by DylanGuedes**
Introduce the usage of Go new ' WithCause' functions to make our `context cancellation` errors more descriptive.
I'm not yet changing every `context.WithCancel` here because, from what I learned, places were we immediately cancel the context won't strongly benefit from this.

### CODE REVIEW

File: pkg/querier/worker/scheduler_processor.go [hunk 1]
```diff
@@ -89,7 +90,7 @@ func (sp *schedulerProcessor) processQueriesOnSingleStream(workerCtx context.Con
 	// Run the querier loop (and so all the queries) in a dedicated context that we call the "execution context".
 	// The execution context is cancelled once the workerCtx is cancelled AND there's no inflight query executing.
 	execCtx, execCancel, inflightQuery := newExecutionContext(workerCtx, sp.log)
-	defer execCancel()
+	defer execCancel(errors.New("scheduler processor execution context canceled"))
```
Comment (cstyan) [inline pkg/querier/worker/scheduler_processor.go:93]: should we include more context here? the comment above makes it pretty clear when we would be cancelling the execution context
```
// The execution context is cancelled once the workerCtx is cancelled AND there's no inflight query executing.
```

**Review by cstyan — COMMENTED** [review]
these errors are not propagated back to end users, right? except maybe query timeouts?

File: pkg/querier/worker/scheduler_processor.go (hunk 1 above)
Comment (DylanGuedes) [reply pkg/querier/worker/scheduler_processor.go:93]: > should we include more context here? the comment above makes it pretty clear when we would be cancelling the execution context

yep it is pretty much what is in the comment, but the comment won't be present on the log line so you won't have access to it when querying the logs.

**Comment by DylanGuedes** [comment]
> these errors are not propagated back to end users, right? except maybe query timeouts?

yep, I'm expecting the query timeouts to be propagated back but the others won't happen

**Review by cstyan — APPROVED** [review]

**Pull request merged**